File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AntiAFK.java
package me.zeroeightsix.kami.module.modules.misc;

import baritone.api.BaritoneAPI;
import baritone.api.pathing.goals.GoalXZ;
import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.network.play.client.CPacketAnimation;
import net.minecraft.network.play.server.SPacketChat;
import net.minecraft.util.EnumHand;

import java.util.Objects;
import java.util.Random;

import static me.zeroeightsix.kami.util.MathsUtils.reverseNumber;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendServerMessage;

/**
 * Created by 086 on 16/12/2017.
 * Updated by dominikaaaa on 21/04/20
 * TODO: Path finding to stay inside 1 chunk
 * TODO: Render which chunk is selected
 */
@Module.Info(
        name = "AntiAFK",
        category = Module.Category.MISC,
        description = "Prevents being kicked for AFK"
)
public class AntiAFK extends Module {

    private final Setting<Integer> frequency = register(Settings.integerBuilder("ActionFrequency").withMinimum(1).withMaximum(100).withValue(40).build());
    public Setting<Boolean> autoReply = register(Settings.b("AutoReply", true));
    private final Setting<Boolean> swing = register(Settings.b("Swing", true));
    private final Setting<Boolean> jump = register(Settings.b("Jump", true));
    private final Setting<Boolean> squareWalk = register(Settings.b("SquareWalk", true));
    private final Setting<Integer> radius = register(Settings.integerBuilder("Radius").withMinimum(1).withValue(64).build());
    private final Setting<Boolean> turn = register(Settings.b("Turn", true));

    private final Random random = new Random();

    private final int[] squareStartCoords = {0, 0};
    private int squareStep = 0;

    @Override
    public void onEnable() {
        if (mc.player == null)
            return;

        squareStartCoords[0] = (int) mc.player.posX;
        squareStartCoords[1] = (int) mc.player.posZ;
    }

    @Override
    public void onDisable() {
        if (mc.player == null)
            return;

        if (isBaritoneActive())
            BaritoneAPI.getProvider().getPrimaryBaritone().getPathingBehavior().cancelEverything();
    }

    @Override
    public void onUpdate() {
        if (mc.playerController.getIsHittingBlock()) return;

        if (swing.getValue() && mc.player.ticksExisted % (0.5 * getFrequency()) == 0) {
            Objects.requireNonNull(mc.getConnection()).sendPacket(new CPacketAnimation(EnumHand.MAIN_HAND));
        }

        if (squareWalk.getValue() && mc.player.ticksExisted % getFrequency() == 0 && !isBaritoneActive()) {
            int r = clamp(radius.getValue());
            switch (squareStep) {
                // +z
                case 0:
                    baritoneGotoXZ(squareStartCoords[0], squareStartCoords[1] + r);
                    break;
                // +x
                case 1:
                    baritoneGotoXZ(squareStartCoords[0] + r, squareStartCoords[1] + r);
                    break;
                // -z
                case 2:
                    baritoneGotoXZ(squareStartCoords[0] + r, squareStartCoords[1]);
                    break;
                // -x
                case 3:
                    baritoneGotoXZ(squareStartCoords[0], squareStartCoords[1]);
                    break;
            }
            squareStep = (squareStep + 1) % 4;
        }

        if (jump.getValue() && mc.player.ticksExisted % (2 * getFrequency()) == 0) {
            mc.player.jump();
        }

        if (turn.getValue() && mc.player.ticksExisted % (0.375 * getFrequency()) == 0) {
            mc.player.rotationYaw = random.nextInt(360) - makeNegRandom(180);
        }
    }

    @EventHandler
    public Listener<PacketEvent.Receive> receiveListener = new Listener<>(event -> {
        if (autoReply.getValue() && event.getPacket() instanceof SPacketChat && ((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains("whispers: ") && !((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains(mc.player.getName())) {
            sendServerMessage("/r I am currently AFK and using KAMI Blue!");
        }
    });

    private boolean isBaritoneActive() {
        return BaritoneAPI.getProvider().getPrimaryBaritone().getCustomGoalProcess().isActive();
    }

    private void baritoneGotoXZ(int x, int z) {
        BaritoneAPI.getProvider().getPrimaryBaritone().getCustomGoalProcess().setGoalAndPath(new GoalXZ(x, z));
    }

    private int clamp(int val) {
        if (val < 0)
            return 0;
        return val;
    }

    private float getFrequency() {
        return reverseNumber(frequency.getValue(), 1, 100);
    }

    private int makeNegRandom(int input) {
        int rand = random.nextBoolean() ? 1 : 0;
        if (rand == 0) {
            return -input;
        } else {
            return input;
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AntiAFK.kt
package me.zeroeightsix.kami.module.modules.misc

import baritone.api.BaritoneAPI
import baritone.api.pathing.goals.GoalXZ
import me.zeroeightsix.kami.event.events.BaritoneSettingsInitEvent
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.event.events.SafeTickEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Setting
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.BaritoneUtils
import me.zeroeightsix.kami.util.TimerUtils
import me.zeroeightsix.kami.util.event.listener
import me.zeroeightsix.kami.util.text.MessageDetectionHelper
import me.zeroeightsix.kami.util.text.MessageSendHelper.sendServerMessage
import net.minecraft.network.play.server.SPacketChat
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.gameevent.InputEvent
import kotlin.math.max
import kotlin.random.Random

/**
 * TODO: Path finding to stay inside 1 chunk
 * TODO: Render which chunk is selected
 */
@Module.Info(
        name = "AntiAFK",
        category = Module.Category.MISC,
        description = "Prevents being kicked for AFK"
)
object AntiAFK : Module() {
    private val delay = register(Settings.integerBuilder("ActionDelay").withValue(50).withRange(0, 100))
    private val variation = register(Settings.integerBuilder("Variation").withValue(25).withRange(0, 50))
    private val autoReply = register(Settings.b("AutoReply", true))
    private val swing = register(Settings.b("Swing", true))
    private val jump = register(Settings.b("Jump", true))
    private val turn = register(Settings.b("Turn", true))
    private val walk = register(Settings.b("Walk", true))
    private val radius = register(Settings.integerBuilder("Radius").withValue(64).withRange(1, 128))
    private val inputTimeout = register(Settings.integerBuilder("InputTimeout(m)").withValue(0).withRange(0, 15))

    private var startPos: BlockPos? = null
    private var nextActionTick = 0
    private var squareStep = 0
    private var baritoneDisconnectOnArrival = false
    private val inputTimer = TimerUtils.TickTimer(TimerUtils.TimeUnit.MINUTES)

    override fun getHudInfo(): String? {
        return ((System.currentTimeMillis() - inputTimer.time) / 1000L).toString()
    }

    override fun onEnable() {
        inputTimer.reset()
        baritoneDisconnectOnArrival()
    }

    override fun onDisable() {
        startPos = null
        BaritoneUtils.settings()?.disconnectOnArrival?.value = baritoneDisconnectOnArrival
        BaritoneAPI.getProvider().primaryBaritone.pathingBehavior.cancelEverything()
    }

    init {
        listener<BaritoneSettingsInitEvent> {
            baritoneDisconnectOnArrival()
        }

        listener<PacketEvent.Receive> {
            if (!autoReply.value || it.packet !is SPacketChat) return@listener
            if (MessageDetectionHelper.isDirect(true, it.packet.getChatComponent().unformattedText)) {
                sendServerMessage("/r I am currently AFK and using KAMI Blue!")
            }
        }

        listener<InputEvent.MouseInputEvent> {
            if (inputTimeout.value != 0 && isInputting()) {
                inputTimer.reset()
            }
        }

        listener<InputEvent.KeyInputEvent> {
            if (inputTimeout.value != 0 && isInputting()) {
                inputTimer.reset()
            }
        }

        listener<SafeTickEvent> {
            if (inputTimeout.value != 0) {
                if (BaritoneAPI.getProvider().primaryBaritone.customGoalProcess.isActive) inputTimer.reset()
                if (!inputTimer.tick(inputTimeout.value.toLong(), false)) {
                    startPos = null
                    return@listener
                }
            }

            if (mc.player.ticksExisted >= nextActionTick) {
                val random = if (variation.value > 0) (0..variation.value).random() else 0
                nextActionTick = mc.player.ticksExisted + max(delay.value, 1) + random

                when ((getAction())) {
                    Action.SWING -> mc.player.swingArm(EnumHand.MAIN_HAND)
                    Action.JUMP -> mc.player.jump()
                    Action.TURN -> mc.player.rotationYaw = Random.nextDouble(-180.0, 180.0).toFloat()
                }

                if (walk.value && !BaritoneAPI.getProvider().primaryBaritone.customGoalProcess.isActive) {
                    if (startPos == null) startPos = mc.player.position
                    startPos?.let {
                        when (squareStep) {
                            0 -> baritoneGotoXZ(it.x, it.z + radius.value)
                            1 -> baritoneGotoXZ(it.x + radius.value, it.z + radius.value)
                            2 -> baritoneGotoXZ(it.x + radius.value, it.z)
                            3 -> baritoneGotoXZ(it.x, it.z)
                        }
                    }
                    squareStep = (squareStep + 1) % 4
                }
            }
        }
    }

    private fun baritoneDisconnectOnArrival() {
        BaritoneUtils.settings()?.disconnectOnArrival?.let {
            baritoneDisconnectOnArrival = it.value
            it.value = false
        }
    }

    private fun isInputting(): Boolean {
        return (mc.gameSettings.keyBindAttack.isKeyDown
                || mc.gameSettings.keyBindUseItem.isKeyDown
                || mc.gameSettings.keyBindJump.isKeyDown
                || mc.gameSettings.keyBindSneak.isKeyDown
                || mc.gameSettings.keyBindForward.isKeyDown
                || mc.gameSettings.keyBindBack.isKeyDown
                || mc.gameSettings.keyBindLeft.isKeyDown
                || mc.gameSettings.keyBindRight.isKeyDown)
    }

    private fun getAction(): Action? {
        if (!swing.value && !jump.value && !turn.value) return null
        val action = Action.values().random()
        return if (action.setting.value) action else getAction()
    }

    private enum class Action(val setting: Setting<Boolean>) {
        SWING(swing),
        JUMP(jump),
        TURN(turn)
    }

    private fun baritoneGotoXZ(x: Int, z: Int) {
        BaritoneAPI.getProvider().primaryBaritone.customGoalProcess.setGoalAndPath(GoalXZ(x, z))
    }

    init {
        walk.settingListener = Setting.SettingListeners {
            if (BaritoneUtils.settingsInitialized && BaritoneAPI.getProvider().primaryBaritone.customGoalProcess.isActive) {
                BaritoneAPI.getProvider().primaryBaritone.pathingBehavior.cancelEverything()
            }
        }
    }
}